In [2]:
import pandas as pd #makes necessary imports, pandas and numpy
import numpy as np

In [3]:
df = pd.read_csv("https://raw.githubusercontent.com/WalePhenomenon/climate_change/master/fuel_ferc1.csv") #reads csv file from link

In [4]:
df.head() # shows first five rows..preview sake

record_id  utility_id_ferc1  report_year  \
0   f1_fuel_1994_12_1_0_7                 1         1994   
1  f1_fuel_1994_12_1_0_10                 1         1994   
2   f1_fuel_1994_12_2_0_1                 2         1994   
3   f1_fuel_1994_12_2_0_7                 2         1994   
4  f1_fuel_1994_12_2_0_10                 2         1994   

       plant_name_ferc1 fuel_type_code_pudl fuel_unit  fuel_qty_burned  \
0              rockport                coal       ton        5377489.0   
1  rockport total plant                coal       ton       10486945.0   
2                gorgas                coal       ton        2978683.0   
3                 barry                coal       ton        3739484.0   
4             chickasaw                 gas       mcf          40533.0   

   fuel_mmbtu_per_unit  fuel_cost_per_unit_burned  \
0               16.590                      18.59   
1               16.592                      18.58   
2               24.130                      39.72   
3               23.950                      47.21   
4                1.000                       2.77   

   fuel_cost_per_unit_delivered  fuel_cost_per_mmbtu  
0                         18.53                1.121  
1                         18.53                1.120  
2                         38.12                1.650  
3                         45.99                1.970  
4                          2.77                2.570

In [5]:
# QUESTION THREE

In [6]:
df_fuel_type=df.groupby("fuel_type_code_pudl") # groups the dataframe by fuel_type_code_pudl...this creates a groupby object

In [7]:
df_fuel_type.mean()["fuel_cost_per_unit_burned"].sort_values(ascending=False) # displays the average of the 'fuel_cost_per_unit_burned' instances corresponding to the fuel types and sorts them in descending order 

fuel_type_code_pudl
nuclear    4955.157002
oil         168.877086
coal         67.421830
waste        19.518122
other        18.253856
gas          13.659397
Name: fuel_cost_per_unit_burned, dtype: float64

The answer is gas

In [8]:
# QUESTION FOUR

In [10]:
df.describe().loc[["std","75%"],["fuel_mmbtu_per_unit"]].round(decimals = 2) # this selects standard deviation and fuel_mmbtu_per_unit from the describe function dataframe

fuel_mmbtu_per_unit
std                10.60
75%                17.01

Standard deviation : 10.60 ;   75%ile : 17.01

In [11]:
# QUESTION FIVE

In [12]:
df["fuel_qty_burned"].skew() #SKEWNESS

15.851495469109503

In [13]:
df["fuel_qty_burned"].kurt() #KURTOSIS

651.3694501337732

In [14]:
# QUESTION SIX

In [15]:
df.isnull().sum() # checks for all instances of a null value and counts all the instances

record_id                         0
utility_id_ferc1                  0
report_year                       0
plant_name_ferc1                  0
fuel_type_code_pudl               0
fuel_unit                       180
fuel_qty_burned                   0
fuel_mmbtu_per_unit               0
fuel_cost_per_unit_burned         0
fuel_cost_per_unit_delivered      0
fuel_cost_per_mmbtu               0
dtype: int64

Fuel unit is the feature with missing values

In [16]:
df.count()[0] #Total number of rows

29523

In [17]:
(df.isnull().sum().loc["fuel_unit"] /df.count()[0] *100).round(3)# To calculate percentage of missing rows.

0.61

In [18]:
# QUESTION SEVEN

In [19]:
type(df["fuel_unit"][0]) # To get type of value pesent in fuel_unit column

str

Therefore the value type is categorical and mode(highest frequency) imputation can be used

In [20]:
# QUESTION EIGHT

In [21]:
df.corr() #get correlation matrix

utility_id_ferc1  report_year  fuel_qty_burned  \
utility_id_ferc1                      1.000000     0.093323        -0.057447   
report_year                           0.093323     1.000000         0.012952   
fuel_qty_burned                      -0.057447     0.012952         1.000000   
fuel_mmbtu_per_unit                  -0.066946    -0.110853        -0.080946   
fuel_cost_per_unit_burned            -0.037863     0.013599        -0.018535   
fuel_cost_per_unit_delivered         -0.016414    -0.014043        -0.003551   
fuel_cost_per_mmbtu                   0.006122     0.010261        -0.001896   

                              fuel_mmbtu_per_unit  fuel_cost_per_unit_burned  \
utility_id_ferc1                        -0.066946                  -0.037863   
report_year                             -0.110853                   0.013599   
fuel_qty_burned                         -0.080946                  -0.018535   
fuel_mmbtu_per_unit                      1.000000                  -0.010034   
fuel_cost_per_unit_burned               -0.010034                   1.000000   
fuel_cost_per_unit_delivered            -0.009039                   0.011007   
fuel_cost_per_mmbtu                     -0.005884                  -0.000437   

                              fuel_cost_per_unit_delivered  \
utility_id_ferc1                                 -0.016414   
report_year                                      -0.014043   
fuel_qty_burned                                  -0.003551   
fuel_mmbtu_per_unit                              -0.009039   
fuel_cost_per_unit_burned                         0.011007   
fuel_cost_per_unit_delivered                      1.000000   
fuel_cost_per_mmbtu                              -0.000109   

                              fuel_cost_per_mmbtu  
utility_id_ferc1                         0.006122  
report_year                              0.010261  
fuel_qty_burned                         -0.001896  
fuel_mmbtu_per_unit                     -0.005884  
fuel_cost_per_unit_burned               -0.000437  
fuel_cost_per_unit_delivered            -0.000109  
fuel_cost_per_mmbtu                      1.000000

In [24]:
sec_third_lowest = np.abs(df.corr()).loc["fuel_cost_per_unit_burned"].sort_values(ascending=False) # selects the fuel_cost_per_unit_burned column and sort the absolute values in descending order
sec_third_lowest

fuel_cost_per_unit_burned       1.000000
utility_id_ferc1                0.037863
fuel_qty_burned                 0.018535
report_year                     0.013599
fuel_cost_per_unit_delivered    0.011007
fuel_mmbtu_per_unit             0.010034
fuel_cost_per_mmbtu             0.000437
Name: fuel_cost_per_unit_burned, dtype: float64

In [23]:
sec_third_lowest[-3:-1] # selects the second and third to the last values in the above pandas series

fuel_cost_per_unit_delivered    0.011007
fuel_mmbtu_per_unit             0.010034
Name: fuel_cost_per_unit_burned, dtype: float64

Therefore, the features with the 2nd and 3rd lowest correlations with 'fuel_cost_per_unit_burned' are 'fuel_mmbtu_per_unit' and 'fuel_cost_per_unit_delivered' respectively.

In [ ]:
#QUESTION NINE

In [71]:
df_coal = df[df["fuel_type_code_pudl"] == "coal"]  # selects only rows with fuel_type_code of coal

In [159]:
coal_year_group = df_coal.groupby("report_year") #...then groups the dataframe derived above by report_year

In [161]:
fcpub = coal_year_group.mean()["fuel_cost_per_unit_burned"] #displays the average of the 'fuel_cost_per_unit_burned' instances corresponding to the report_years  

In [162]:
fcpub.head()

report_year
1994     31.546467
1995     30.677442
1996     29.454299
1997    104.897309
1998     27.616234
Name: fuel_cost_per_unit_burned, dtype: float64

In [164]:
( fcpub.loc[1994] - fcpub.loc[1998] ) / fcpub.loc[1994] * 100 # calculates percentage difference between 1994 and 1998 values

12.45855196328015

Therefore, the percentage change in fuel_cost_per_unit_burned in 1998 compared to 1994 is 12.46%

In [ ]:
#QUESTION TEN

In [87]:
year_group = df.groupby("report_year") # groups the main dataframe by report_year

In [152]:
year_group_fcpud = year_group.mean()["fuel_cost_per_unit_delivered"].sort_values(ascending=False) #displays the average of the 'fuel_cost_per_unit_delivered' instances corresponding to the report_years and sorts them in descending order  
year_group_fcpud.head()

report_year
2018    1204.041209
1998     716.015659
2010     184.509668
2000     173.192467
1994     124.663939
Name: fuel_cost_per_unit_delivered, dtype: float64

Therefore, the year with the highest average fuel cost per unit delivered is 2018